### PR TITLE
ONNX export: propagate node metadata across passes

### LINKED ISSUE #45255: ONNX: source range and node scopes do not appear in final graph
## 🐛 Bug

When exporting a traced or scripted module, some calls (such as aten expansions) and passes do not carry node metadata forward. Source information of the nodes in the exported graph then becomes sparse.

## To Reproduce

1. Run the code below with `torch.onnx.export(..., strip_doc_string=False)` (if necessary, I can construct a simpler example)
1. Observe node docstrings in exported ONNX graphs

```python
class SecondModule(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, tensor):
        return tensor + tensor


class RecursiveModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.val = torch.nn.Parameter(torch.tensor([2.]))
        self.mod1 = SecondModule()
        self.mod2 = SecondModule()
        self.ln = torch.nn.LayerNorm([1, ])

    def forward(self, input):
        return self.ln(self.mod2(self.val * self.mod1(input)))

mymod = torch.jit.script(RecursiveModule())
```

## Expected behavior

Exporting should have all nodes augmented with `SourceRange` information. Instead, only the pure operations have the relevant information attached to them.

## Environment

Platform-independent.

### PR DESCRIPTION
**Pull request by tbennun**
Fixes #45255 

Mostly straightforward. Only downside in this PR is the lack of more scalable way to check for all newly-created nodes in `callPySymbolicFunction`. The other options were:
* Create a scope within the node's scope and loop through all nodes that correspond to the scope. The code would still need to loop through all nodes.
* Add extra state to the graph (no good reason to do so).
* Add extra state to the ONNX exporter, since python calls go back to `g.op(...)` (no good reason to do so, also not very pythonic).

cc @BowenBao @neginraoof

### CODE REVIEW

File: torch/csrc/jit/passes/onnx.cpp [hunk 1]
```diff
@@ -294,15 +294,27 @@ void BlockToONNX(
       py_inputs[input_nr++] = py::cast(envFn(input));
     }
 
+    Graph* g = ctx.block->owningGraph();
+    std::unordered_set<Node*> nodes_before;
+    for (auto node : g->nodes())
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx.cpp:299]: please add curly braces around loop body

**Review by garymm — CHANGES_REQUESTED** [review]
@tbennun thanks for this change, it's awesome!

Could you please add a unit test?
The sourceRange appears to be exposed to Python so you could add it to test/jit/test_onnx_export.py

Please also rebase.

**Comment by garymm** [comment]
@shubhambhokare1 CC

**Comment by tbennun** [comment]
Thanks for reviewing! Will take care of it ASAP.

**Comment by tbennun** [comment]
@garymm yes, apologies for the delay. Due to a deadline this Friday, I'll take care of it by the beginning of the next week (the 12th EOD). I already rebased and performed the change in the comment but still need to add the unit test.

**Review by garymm — CHANGES_REQUESTED** [review]
Thanks! Once you've made the minor changes I think this is ready to go, but we'll need a Facebook employee to actually merge it.

**Review by garymm — APPROVED** [review]

**Comment by garymm** [comment]
@msaroufim please import this community contribution.

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**